module/tactical/tactical_class.py

```diff
@@ -175,6 +185,72 @@ def _tactical_books_get(self, skip_first_screenshot=True):
         logger.warning('No book found.')
         raise ScriptError('No book found, after 15 attempts.')
 
+    def _tactical_selected_get(self):
+        """
+        Find the selected book onscreen
+        Called after _tactical_books_get
+        """
+        selected = None
+        for book in self.books:
+            if book.check_selected(self.device.image):
+                selected = book
+                break
+        return selected
+
+    def _tactical_books_filter_exp(self):
+        """
+        Complex filter to remove specific grade
+        books from self.books based on current
+        progress of the tactical skill.
+        """
+        selected = self._tactical_selected_get()
+        if selected is None:
+            return
+
+        current, remain, total = SKILL_EXP.ocr(self.device.image)
+        if total == 5800:
+            # Read 'current' and 'remain' are inaccurate
+            # as selected exp_value is factored into it
+            current -= selected.exp_value
+            remain += selected.exp_value
+            logger.info('About to reach level 10; will remove '
+                        'detected books based on actual '
+                       f'progress: {current}/{total}; {remain}')
+
+            reverse_tier = reversed(list(selected.exp_tier.keys()))
+            backup = remain
+            for tier in reverse_tier:
+                if not tier:
+                    continue
+
+                # Get the corresponding exp, factor based on tier
+                exp_value = selected.exp_tier[tier]
+                factor = 1.5 if tier < 4 else 2
+
+                # Reset helper variables
+                groups = []
+                overflow = 0
+                remain = backup
+
+                # Modify 'remain' if overflow allowed
+                if self.config.ControlExpOverflow_Enable:
+                    overflow = getattr(self.config, f'ControlExpOverflow_T{tier}Allow')
+                    logger.info(f'T{tier} overflow allowed: {overflow}')
+                    remain += overflow
+
+                # Determine applicable books for removal
+                # Retain at least non-bonus T1 books if
+                # nothing else
+                if remain < (exp_value * factor):
+                    groups.append(self.books.select(tier=tier, exp=True))
+                if exp_value != 100 and remain < exp_value:
+                    groups.append(self.books.select(tier=tier, exp=False))
+                logger.info(f'T{tier} removing: {sum([group.count for group in groups])}')
+
+                # Remove applicable books to prevent waste
+                for group in groups:
+                    self.books = self.books.delete(group)
```

Comment:
Each `Book` object has attribute `exp_value`, no need to calculate again.
A simple implement would be:
```python
def filter_func(book):
    if current + book.exp_value > total + self.config.ControlExpOverflow_T*Allow:
        return False
    return True

self.books = SelectedGrids([b for b in self.books if filter_func(b)])
```

Reply:
Thanks for the sample, quite elegant. I'll run simulations to ensure its behaving as intended. From what it looks like we don't need the `ControlExpOverflow_Enable` then, people can just put in zero if they rather not have any overflow for whatever tier.

Separate issue; when going from level 8 to 9, the OCR will read a digit count X/5800, that makes keeping track of `current` more difficult. 
```
2021-11-26 11:15:01.999 | INFO | [OCR_SKILL_EXP 0.082s] 0/5800
2021-11-26 11:15:02.000 | INFO | About to reach level 10; will remove detected books based on actual progress: -3000/5800; 8800
```
Thinking of capturing an asset where it might say lv 8+1, might be a while since I don't have any at the moment.  Using lv 9 asset + 5800 might work too...

Now that I think about it, also might have been read as `0/5800` because could print as `MAX/5800`. Hmm, then wouldn't be able to gauge an accurate `current` can't assume you didn't overflow already in this case and wouldn't know how much overflowed. Guess I can see the reason this filter being more difficult than I anticipated.

Still managed process fine though for now.

Reply:
Try this pseudo code:
```
(do OCR)
if current == 0:
    # Lv9+1, using the first book will level to max
    (select the last book)
    (do OCR)
    if current == 0:
        # Lv9+1, using the book with minimum exp will still level to max
        (current = total - selected_book.exp_value)
    else:
        # Lv9, current < total, able to get actual current exp
        (current -= selected_book.exp_value)
else:
    # Lv9, all good, able to calculate
    (current -= selected_book.exp_value)
```

Reply:
Definitely was overthinking and I only just realized I was mistaken about the `Lv8+1`, it does not change `total` dynamically like I had thought.  So only have to worry about 1 case which your pseudo code did simplify a whole lot.

Did a live test since using a fixed image wouldn't have worked.  Fortunately had one skill close to max level and seems to work
```
2021-11-27 12:05:27.895 | INFO | [Book_count] 12
2021-11-27 12:05:27.895 | INFO | [Books] [Red_T4_Exp, Blue_T4, Yellow_T4, Red_T3
_Exp, Blue_T3, Yellow_T3, Red_T2_Exp, Blue_T2, Yellow_T2, Red_T1_Exp, Blue_T1, Y
ellow_T1]
screenshot: 0.4775776863 s
2021-11-27 12:05:28.380 | INFO | [Book_count] 12
2021-11-27 12:05:28.380 | INFO | [Books] [Red_T4_Exp, Blue_T4, Yellow_T4, Red_T3
_Exp, Blue_T3, Yellow_T3, Red_T2_Exp, Blue_T2, Yellow_T2, Red_T1_Exp, Blue_T1, Y
ellow_T1]
2021-11-27 12:05:28.381 | INFO | Click ( 280,  326) @ BOOKS_GRID_0_0
2021-11-27 12:05:28.381 | INFO | <<< MINITOUCH INIT >>>
2021-11-27 12:05:30.383 | INFO | Minitouch bind to port 20793
2021-11-27 12:05:30.513 | INFO | Minitouch forward port 20793
2021-11-27 12:05:30.517 | INFO | minitouch running on port: 20793, pid: 5185
2021-11-27 12:05:30.517 | INFO | max_contact: 2; max_x: 32767; max_y: 32767; max
_pressure: 0
screenshot: 0.5068545341 s
2021-11-27 12:05:31.472 | INFO | Loading OCR model: ./bin/cnocr_models/azur_lane

2021-11-27 12:05:31.549 | INFO | [OCR_SKILL_EXP 0.090s] 0/5800
2021-11-27 12:05:31.549 | INFO | Click ( 979,  483) @ BOOKS_GRID_5_1
screenshot: 0.5137686729 s
2021-11-27 12:05:32.543 | INFO | [OCR_SKILL_EXP 0.023s] 5450/5800
2021-11-27 12:05:32.543 | INFO | About to reach level 10; will remove detected b
ooks based on actual progress: 5350/5800; 450
2021-11-27 12:05:32.544 | INFO | [EXP Filtered] 6
2021-11-27 12:05:32.544 | INFO | [Books] [Red_T2_Exp, Blue_T2, Yellow_T2, Red_T1
_Exp, Blue_T1, Yellow_T1]
2021-11-27 12:05:32.544 | INFO | [Book_sort] Red_T2_Exp > Red_T1_Exp > Blue_T2 >
 Yellow_T2 > Blue_T1 > Yellow_T1 > first
2021-11-27 12:05:32.545 | INFO | Click ( 286,  475) @ BOOKS_GRID_0_1.
```